File: projects/PROJ-046-the-relationship-between-sleep-chronotyp/tests/test_report.py
"""
Unit tests for report structure validation (Task T025).

This module validates that the generated R-Markdown report contains all
required sections as specified in the project requirements:
- Descriptive statistics
- ANCOVA results
- Effect sizes (Cohen's d)
- Power analysis (G*Power summary)
- Sensitivity analysis table
- Reliability metrics (Cronbach's alpha)
- Data exclusion summary

The test verifies the existence of these sections by parsing the
generated HTML/Markdown report.
"""
import os
import re
import unittest
from pathlib import Path


class TestReportStructure(unittest.TestCase):
    """Test suite for validating report structure."""

    # Paths relative to project root
    PROJECT_ROOT = Path(__file__).resolve().parent.parent
    REPORT_PATH = PROJECT_ROOT / "reports" / "chronotype_moral_analysis.html"
    SENSITIVITY_CSV = PROJECT_ROOT / "data" / "derived" / "sensitivity_sweep.csv"
    RELIABILITY_CSV = PROJECT_ROOT / "data" / "derived" / "reliability_metrics.csv"
    EXCLUSIONS_LOG = PROJECT_ROOT / "data" / "derived" / "exclusions.log"

    def setUp(self):
        """Set up test fixtures."""
        self.report_path = self.REPORT_PATH
        self.required_sections = [
            r"(?i)Descriptive\s+Statistics",
            r"(?i)ANCOVA\s+Results",
            r"(?i)Effect\s+Sizes",
            r"(?i)Power\s+Analysis",
            r"(?i)Sensitivity\s+Analysis",
            r"(?i)Reliability\s+Metrics",
            r"(?i)Data\s+Exclusion\s+Summary"
        ]

    def test_report_file_exists(self):
        """Verify that the report file exists."""
        self.assertTrue(
            self.report_path.exists(),
            f"Report file not found at {self.report_path}. "
            "Ensure code/04_report.Rmd has been rendered."
        )

    def test_report_contains_required_sections(self):
        """Verify that the report contains all required sections."""
        if not self.report_path.exists():
            self.skipTest("Report file does not exist; skipping section check.")

        with open(self.report_path, "r", encoding="utf-8") as f:
            content = f.read()

        missing_sections = []
        for pattern in self.required_sections:
            if not re.search(pattern, content):
                missing_sections.append(pattern)

        self.assertEqual(
            len(missing_sections), 0,
            f"Report is missing the following sections:\n"
            + "\n".join(missing_sections)
        )

    def test_sensitivity_table_has_required_columns(self):
        """Verify that sensitivity_sweep.csv has required columns."""
        if not self.SENSITIVITY_CSV.exists():
            self.skipTest("Sensitivity sweep CSV not found; skipping column check.")

        # Read first line to check headers
        with open(self.SENSITIVITY_CSV, "r", encoding="utf-8") as f:
            header_line = f.readline().strip()

        required_cols = ["subscale", "alpha_threshold", "p_value", "significant"]
        cols = [c.strip() for c in header_line.split(",")]

        missing_cols = [c for c in required_cols if c not in cols]
        self.assertEqual(
            len(missing_cols), 0,
            f"Sensitivity sweep CSV missing columns: {missing_cols}"
        )

    def test_sensitivity_table_has_multiple_thresholds(self):
        """Verify that sensitivity table includes at least three alpha thresholds."""
        if not self.SENSITIVITY_CSV.exists():
            self.skipTest("Sensitivity sweep CSV not found; skipping threshold count check.")

        import pandas as pd
        df = pd.read_csv(self.SENSITIVITY_CSV)

        unique_thresholds = df["alpha_threshold"].unique()
        self.assertGreaterEqual(
            len(unique_thresholds), 3,
            f"Sensitivity table must include at least 3 alpha thresholds. "
            f"Found: {len(unique_thresholds)}"
        )

    def test_reliability_metrics_exists_and_valid(self):
        """Verify that reliability_metrics.csv exists and has valid alpha values."""
        if not self.RELIABILITY_CSV.exists():
            self.skipTest("Reliability metrics CSV not found; skipping validation.")

        import pandas as pd
        df = pd.read_csv(self.RELIABILITY_CSV)

        required_cols = ["subscale", "cronbach_alpha", "n_items"]
        missing_cols = [c for c in required_cols if c not in df.columns]
        self.assertEqual(
            len(missing_cols), 0,
            f"Reliability metrics CSV missing columns: {missing_cols}"
        )

        # Check that alpha values are between 0 and 1
        self.assertTrue(
            all((df["cronbach_alpha"] >= 0) & (df["cronbach_alpha"] <= 1)),
            "Cronbach's alpha values must be between 0 and 1."
        )

    def test_exclusions_log_exists(self):
        """Verify that exclusions.log exists."""
        self.assertTrue(
            self.EXCLUSIONS_LOG.exists(),
            f"Exclusions log not found at {self.EXCLUSIONS_LOG}. "
            "Ensure T012.5 has been run to generate this file."
        )


if __name__ == "__main__":
    unittest.main()

File: projects/PROJ-046-the-relationship-between-sleep-chronotyp/tests/test-report.R
# tests/test-report.R
# Validation test for report structure (US3)
# Task: T025
#
# This test verifies that the generated report (HTML/PDF) contains all
# required sections as specified in the feature requirements (SC-003).
#
# Required sections to verify:
# 1. Descriptive Statistics Table
# 2. ANCOVA Results (with Bonferroni correction)
# 3. Effect Sizes (Cohen's d)
# 4. Power Analysis Summary
# 5. Sensitivity Analysis (Sweep results)
# 6. Cronbach's Alpha Reliability Metrics
# 7. Data Exclusion Summary
#
# The test reads the rendered report file and checks for the presence
# of these sections. It fails if any required section is missing.

library(testthat)
library(rmarkdown)
library(purrr)
library(stringr)

# Helper: Get project root
get_project_root <- function() {
  # Assume running from project root or tests/
  if (file.exists("code/00_config.R")) {
    return(".")
  } else if (file.exists("../code/00_config.R")) {
    return("..")
  } else {
    # Fallback: look for .git or specific known file
    path <- getwd()
    while (length(path) > 1 && !file.exists(file.path(path, ".git"))) {
      path <- dirname(path)
    }
    return(path)
  }
}

# Helper: Ensure report exists before testing
ensure_report_exists <- function(report_path) {
  if (!file.exists(report_path)) {
    stop(paste("Report file not found:", report_path,
               ". Please run code/04_report.Rmd first."))
  }
}

# Helper: Read report content (HTML or PDF text extraction)
read_report_content <- function(report_path) {
  ext <- tools::file_ext(report_path)

  if (ext == "html") {
    # Read HTML as text
    content <- readLines(report_path, warn = FALSE)
    return(paste(content, collapse = "\n"))
  } else if (ext == "pdf") {
    # For PDF, we need to extract text.
    # If pdftools is not available, we skip PDF text extraction
    # and just check file existence (less robust but avoids dependency).
    if (requireNamespace("pdftools", quietly = TRUE)) {
      text <- pdftools::pdf_text(report_path)
      return(paste(text, collapse = "\n"))
    } else {
      warning("pdftools not available; cannot extract text from PDF report.")
      return("")
    }
  } else {
    stop(paste("Unsupported report format:", ext))
  }
}

# Helper: Check if a section header exists in the content
# Handles common Markdown/HTML headers: # Section, <h1>Section</h1>
has_section <- function(content, section_name) {
  # Pattern for Markdown headers: # Section Name
  pattern_md <- paste0("^\\s*#\\s+", section_name, "\\s*$")
  # Pattern for HTML headers: <h1>Section Name</h1> (case insensitive)
  pattern_html <- paste0("<h[12][^>]*>\\s*", section_name, "\\s*</h[12]>")

  # Check both patterns (case insensitive)
  md_match <- str_detect(content, regex(pattern_md, ignore_case = TRUE))
  html_match <- str_detect(content, regex(pattern_html, ignore_case = TRUE))

  return(md_match || html_match)
}

# Test Suite
describe("Report Structure Validation (T025)", {

  it("should verify the report file exists", {
    root <- get_project_root()
    report_path <- file.path(root, "reports", "chronotype_moral_analysis.html")

    # If HTML doesn't exist, try PDF
    if (!file.exists(report_path)) {
      report_path <- file.path(root, "reports", "chronotype_moral_analysis.pdf")
    }

    ensure_report_exists(report_path)
    expect_true(file.exists(report_path))
  })

  it("should contain 'Descriptive Statistics' section", {
    root <- get_project_root()
    report_path <- file.path(root, "reports", "chronotype_moral_analysis.html")
    if (!file.exists(report_path)) {
      report_path <- file.path(root, "reports", "chronotype_moral_analysis.pdf")
    }
    ensure_report_exists(report_path)

    content <- read_report_content(report_path)
    expect_true(has_section(content, "Descriptive Statistics"),
                info = "Missing 'Descriptive Statistics' section")
  })

  it("should contain 'ANCOVA Results' section", {
    root <- get_project_root()
    report_path <- file.path(root, "reports", "chronotype_moral_analysis.html")
    if (!file.exists(report_path)) {
      report_path <- file.path(root, "reports", "chronotype_moral_analysis.pdf")
    }
    ensure_report_exists(report_path)

    content <- read_report_content(report_path)
    expect_true(has_section(content, "ANCOVA Results"),
                info = "Missing 'ANCOVA Results' section")
  })

  it("should contain 'Effect Sizes' section", {
    root <- get_project_root()
    report_path <- file.path(root, "reports", "chronotype_moral_analysis.html")
    if (!file.exists(report_path)) {
      report_path <- file.path(root, "reports", "chronotype_moral_analysis.pdf")
    }
    ensure_report_exists(report_path)

    content <- read_report_content(report_path)
    expect_true(has_section(content, "Effect Sizes"),
                info = "Missing 'Effect Sizes' section")
  })

  it("should contain 'Power Analysis' section", {
    root <- get_project_root()
    report_path <- file.path(root, "reports", "chronotype_moral_analysis.html")
    if (!file.exists(report_path)) {
      report_path <- file.path(root, "reports", "chronotype_moral_analysis.pdf")
    }
    ensure_report_exists(report_path)

    content <- read_report_content(report_path)
    expect_true(has_section(content, "Power Analysis"),
                info = "Missing 'Power Analysis' section")
  })

  it("should contain 'Sensitivity Analysis' section", {
    root <- get_project_root()
    report_path <- file.path(root, "reports", "chronotype_moral_analysis.html")
    if (!file.exists(report_path)) {
      report_path <- file.path(root, "reports", "chronotype_moral_analysis.pdf")
    }
    ensure_report_exists(report_path)

    content <- read_report_content(report_path)
    expect_true(has_section(content, "Sensitivity Analysis"),
                info = "Missing 'Sensitivity Analysis' section")
  })

  it("should contain 'Reliability Metrics' section", {
    root <- get_project_root()
    report_path <- file.path(root, "reports", "chronotype_moral_analysis.html")
    if (!file.exists(report_path)) {
      report_path <- file.path(root, "reports", "chronotype_moral_analysis.pdf")
    }
    ensure_report_exists(report_path)

    content <- read_report_content(report_path)
    expect_true(has_section(content, "Reliability Metrics"),
                info = "Missing 'Reliability Metrics' section (Cronbach's alpha)")
  })

  it("should contain 'Data Exclusion Summary' section", {
    root <- get_project_root()
    report_path <- file.path(root, "reports", "chronotype_moral_analysis.html")
    if (!file.exists(report_path)) {
      report_path <- file.path(root, "reports", "chronotype_moral_analysis.pdf")
    }
    ensure_report_exists(report_path)

    content <- read_report_content(report_path)
    expect_true(has_section(content, "Data Exclusion Summary"),
                info = "Missing 'Data Exclusion Summary' section")
  })

  it("should contain at least three alpha thresholds in sensitivity table", {
    root <- get_project_root()
    report_path <- file.path(root, "reports", "chronotype_moral_analysis.html")
    if (!file.exists(report_path)) {
      report_path <- file.path(root, "reports", "chronotype_moral_analysis.pdf")
    }
    ensure_report_exists(report_path)

    content <- read_report_content(report_path)

    # Check for the specific alpha values mentioned in T027: 0.01, 0.0125, 0.015
    # We look for these numbers in the context of the sensitivity section
    # A simple check for the presence of these strings is a proxy for the table
    has_01 <- str_detect(content, "0\\.01")
    has_0125 <- str_detect(content, "0\\.0125")
    has_015 <- str_detect(content, "0\\.015")

    # We expect at least two of them to be present if the table is rendered correctly
    # (0.01 is the main threshold, others are sweep)
    expect_true(sum(c(has_01, has_0125, has_015)) >= 2,
                info = "Sensitivity table should list results for at least three alpha thresholds (0.01, 0.0125, 0.015)")
  })

})
